libnetwork/iptables/firewalld.go

```diff
@@ -159,13 +184,24 @@ func OnReloaded(callback func()) {
 }
 
 // Call some remote method to see whether the service is actually running.
-func checkRunning() bool {
+func checkRunning(ctx context.Context) (bool, error) {
+
+	var err error
 	if connection == nil {
-		return false
+		return false, nil
+	}
+
+	var currentState string
+	err = connection.sysObj.CallWithContext(ctx, "org.freedesktop.DBus.Properties.Get", 0, dbusInterface, "state").Store(&currentState)
+
+	if err != nil {
+		return false, fmt.Errorf("failed to get current state of dbus connection: %w", err)
 	}
-	var zone string
-	err := connection.sysObj.Call(dbusInterface+".getDefaultZone", 0).Store(&zone)
-	return err == nil
+	// Running and failed are treated the same as we only want to verify if Firewalld is up and accepting dbus RPC calls
+	// If Firewalld is in INIT and a timeout has been exceeded dockerd should continue as if firewalld is running
+	// and will eventually finish initialization
+	log.G(ctx).Debugf("Firewalld's current reported state is, %s", currentState)
+	return (currentState == "RUNNING" || currentState == "FAILED" || (currentState == "INIT" && err == context.DeadlineExceeded)), nil
```

Comment:
Have you managed to catch firewalld in the `INIT` state to test this?

I ask because I'm not sure `currentState` will be `INIT` if the timeout is hit ... the dbus responses are likely to be quick, because everything's working normally. So, if firewalld is stuck in `INIT` for a while, the `DeadlineExceeded` might be because the 2.5s timeout had already expired when `checkRunning` was called (so, there probably won't be a dbus request, and `currentState` won't get filled in). Or, it might be that there's not enough time left on the 2.5s timer to get the dbus response.

So, it might need to remember that the last reported state was `INIT`?

It might be simpler to use a normal timer for the 2.5s init timeout, a single context with the 1s timeout for fetching the state, and return the state from `checkRunning`. Then...
- if things are running normally, the current state will always be updated by `checkRunning` (even if the 2.5s timeout is slightly exceeded).
- any error from the dbus call, including `DeadlineExceeded`, can be treated as no-firewalld.
- if the 2.5s timeout expires in `firewalldInit` when the last reported state is `INIT`, startup can just continue without an extra call to `checkRunning`.

Reply:
@robmry 

> Have you managed to catch firewalld in the INIT state to test this?

I simualte INIIT state during testing everytime I make a change and it works just fine as far as I can tell but I will test this further just to be certain. 

> or a while, the DeadlineExceeded might be because the 2.5s timeout had already expired when checkRunning was called (so, there probably won't be a dbus request, and currentState won't get filled in). Or, it might be that there's not enough time left on the 2.5s timer to get the dbus response.
So, it might need to remember that the last reported state was INIT?

That is entirely possible, however, when there is a timeout with no INIT response the previous INIT state isn't relevant as there might have been a state change.  checkRunning() would be reporting back **inaccurate/false** information. The period of time between a state change and the timeout is a really difficult case to catch.  
I believe the current implementation is useful in that it captures one more possible case while not creating an avenue for false negative/positives.

Reply:
> I simualte INIIT state during testing everytime I make a change and it works just fine as far as I can tell but I will test this further just to be certain.

How do you do that - and is it something that could go into a regression test?

Is firewalld still in INIT when the 2.5s timer expires?

> If I follow your suggestion checkRunning() would be reporting back **inaccurate/false** information.

I don't think so ... it might be out-of-date, if INIT completed between the last call to checkRunning and the 2.5s timer expiring. But, that wouldn't make any difference because INIT andRUNNING are treated the same way, apart from a bit of startup delay.

> The period of time between a state change and the timeout is a really difficult case to catch.

Maybe I'm missing something but, I think - if firewalld doesn't get out of INIT before the third attempt to fetch its state (after ~2 seconds), on the next call (at the ~3 second mark, after the 2.5s timer has expired) checkRunning won't even talk to dbus to get the current state. So, `currentState` won't be `"INIT"` and it'll return `false` to say firewalld isn't running - which isn't what we want.

So, there's no need to catch a state change within the timeout - it's no-state-change from INIT that's interesting.

> I believe the current implementation is useful in that it captures one more possible case while not creating an avenue for false negative/positives but proper QA would be the best way to verify this.

What are you thinking of in terms of proper QA? (And is there a reason not to do it?)

Reply:
> How do you do that - and is it something that could go into a regression test?

I stub the responses as INIT until the delayu is reached.

> I don't think so ... it might be out-of-date, if INIT completed between the last call to checkRunning and the 2.5s timer expiring. But, that wouldn't make any difference because INIT andRUNNING are treated the same way, apart from a bit of startup delay.

The case that I observed  was 
**1.)** The previous STATE was INIT
**2.)** The timeout expires
**3.)** Dbus RPC calls are actually in a failed state after the previous successful call


There are several of these cases I saw where the previous state was INIT but later DBUS RPC calls failed.  Its a difficult case to find and a regression test might be flaky. 

>Maybe I'm missing something but, I think - if firewalld doesn't get out of INIT before the third attempt to fetch its state (after ~2 seconds), on the next call (at the ~3 second mark, after the 2.5s timer has expired) checkRunning won't even talk to dbus to get the current state. So, currentState won't be "INIT" and it'll return false to say firewalld isn't running - which isn't what we want.

That's assuming we're not keeping track of the previous state.

> What are you thinking of in terms of proper QA? (And is there a reason not to do it?)
Its not feasible at the moment, but there appear to be complications between dbus, polkit and firewalld that probably need to be tested properly.   I'm not sure how this would work across distributions or if it would be worth it.

Reply:
> I stub the responses as INIT until the delayu is reached.

Is the stub a dbus endpoint, or is the dbus-stuff all stubbed out?

If it's a function replacing dbus ... it needs to check `ctx.Err()` - the real dbus functions won't send a request if it's non-nil. So, the stub should set the result to "INIT" until there's a `ctx.Err()`, then it shouldn't set a result but just return the DeadlineExceeded error.

Reply:
That's exactly what I did for this case. 

I spoke to Brad and he agreed that it might a better to return false on INIT as was suggested in much earlier conversations.  I mentioned the failure cases I've manually tested that were caused by

1.) State is in INIT
2.) The timeout expires
3.) Dbus RPC calls are actually in a failed state after the previous successful call

The consensus was that any timeout should lead to checkRunning() returning false including timeouts where the current state is INIT as we dont know what is going to happen for certain after INIT completes.

Reply:
> That's exactly what I did for this case.

Interesting, I'm not sure how that works.

> The consensus was that any timeout should lead to checkRunning() returning false including timeouts where the current state is INIT as we dont know what is going to happen for certain after INIT completes.

I don't see any special handling in firewalld that'll reject config in state `INIT`.

If there is a way for it to get stuck in INIT, and it's alive enough to report that, the possibilities are:
- it's not updating underlying firewall rules (so, it doesn't matter whether we consider it to be running or not - unless it does start, then we need our rules in place).
- it is updating firewall rules, and we need to configure it.

So, I think we should treat it the same as RUNNING/FAILED, not as "no firewalld" ... that's the current behaviour, changing it without a good reason might be a mistake. (Putting it another way, if we set up firewalld rules when they're not needed, no harm done. If we fail to set them up when they are needed, networking will be mysteriously broken.)

If we do treat RUNNING, FAILED, and INIT the same way ... I think this change can boil down to just a change in the netlink call. Instead of the original call to `.getDefaultZone`, get and log the state (and use a context to get the 1s timeout, just in case dbus/firewalld are unresponsive). There'd be no need to wait in a loop because no different action would be taken, whatever the final state - so the 2.5s timer/context wouldn't be needed.